src/Components/Components/src/Routing/Router.cs

```diff
@@ -93,11 +95,16 @@ public Task SetParametersAsync(ParameterView parameters)
                 throw new InvalidOperationException($"The {nameof(Router)} component requires a value for the parameter {nameof(NotFound)}.");
             }
 
+            if (OnNavigate != null) {
+                NavigationManager.OnNavigate = OnNavigate;
```

Comment:
> What happens if the app has more than one router?

Good question. All of the solutions I'm thinking of this either involve too much cleverness, having a way to map Router instances to OnNavigate handlers, or limiting the feature.

I think we can probably work around this by exposing the `OnNavigate` elsewhere, similar to your though below.

> What should the experience be like if this was configured in a blazor server app?

You can still provide an `OnNavigate` handler but it won't actually do anything since the `RemoteNavigationManager` doesn't provide an override for the `BeforeLocationChangedAsync` property.